### ISSUE TITLE
investigate user reports of "stalled" tests

### ISSUE THREAD

**Issue by critzo** [opentechinstitute/piecewise#159]
Several users of the NDIA instance have reported "stalled" tests. This happens consistently but not every time the test is conducted. Mac, Windows, and Linux browsers have been confirmed to have this issue.

Upon investigation, this issue is confirmed to be a problem with mlab-ns providing clients hostnames for cloud nodes that are offline. The server request is provided with no error, but when a test is attempted, the hostname for the NDT service cannot be resolved.

Tracking in https://github.com/m-lab/mlab-ns/issues/223

https://speed.digitalinclusion.org  

```
Using M-Lab Server ndt-iupui-mlab1-chs0c.measurement-lab.org
main.js:60 Location received
main.js:25 Test started.  Waiting for connection to server...
main.js:25 WebSocket connection to 'wss://ndt-iupui-mlab1-chs0c.measurement-lab.org:3010/ndt_protocol' failed: Error in connection establishment: net::ERR_NAME_NOT_RESOLVED
```

**Comment by critzo** [comment]
Per the issue comments in the mlab-ns repo, this appears to be resolved.

**Comment by critzo** [comment]
Some users still experience stalled tests. Firefox on Ubuntu does not appear work correctly. We think that this may be related to the way Firefox shares location temporarily or permanently. It may be necessary to have a different method of location for Firefox users. @rudietuesdays or @jheretic could you advise on the issues Firefox users are experiencing?

**Comment by critzo** [comment]
Confirmed more info from @nkinkade:

> In firefox. I went to about:config and changed geo.provider.network.url  from "https://www.googleapis.com/geolocation/v1/geolocate?key=%GOOGLE_LOCATION_SERVICE_API_KEY%" to "https://location.services.mozilla.com/v1/geolocate?key=%MOZILLA_API_KEY%" and the test started working.
> It seems that indeed the fundamental issue was the site not being able to obtain a geolocation and just hanging.

**Comment by jheretic** [comment]
This behavior can be generally reproduced by selecting to "Use my browser location" and then denying the browser request (or more generally, having it blocked as is probably more common with Firefox users). The geolocation API call should have a timeout value supplied in order to avoid this in the future.

**Comment by nkinkade** [comment]
@jheretic: I don't _believe_ the browser should have been blocking the request. It's a Firefox profile I have which is configured to pretty much accept anything, do no blocking that Firefox likes to do, but then to delete every single save setting on exit. 

When I clicked on "Use my browser location", Firefox popped  up a dialog asking if I wanted to share my location, to which I selected yes.

It also seems odd to me that it would work with Mozilla's location service, but not Google's, which is the default in Firefox.

This may be user error on my part. I haven't dug into much deeper than what you see in my comment.

**Comment by critzo** [comment]
This has been reported by multiple partners on the NDIA project and confirmed by @nkinkade and myself. I think we need to consider this an interrupt priority to fix now. @laiyi-ohlsen do you agree? We can discuss if needed on today's check in call.

**Comment by laiyi-ohlsen** [comment]
+1

**Comment by laiyi-ohlsen** [comment]
Another report, lmk if contacting the user for more info is useful! 

---------

Was just trying to take your survey which I saw on LinkedIn... kept getting this “wheel of wait” for the speed test.

![unnamed](https://user-images.githubusercontent.com/12607492/77585925-b0100100-6ebb-11ea-8b21-e822a9cc79ed.jpg)



So I ran SpeedTest from my iPad Pro and got this...

![unnamed-1](https://user-images.githubusercontent.com/12607492/77585933-b2725b00-6ebb-11ea-956f-6c2440816794.jpg)


Just went back to your survey page, in Safari, and the wheel of wait is still turning and won’t let me submit my survey.

**Comment by critzo** [comment]
Merged PR #164 and deployed to production.

**Comment by critzo** [comment]
Confirmed in FF and Chrome, this fix does allow users of both browsers to run the test, whether they consent to location or not. However, #162 and #163 are still happening for all submissions. The submissions API endpoint is down and a 500 error is presented in the browser console output.
